test/e2e/network/endpointslice.go

```diff
@@ -299,6 +300,113 @@ var _ = SIGDescribe("EndpointSlice", func() {
 	})
+
+	// Ensure that garbage collection works well when a Service is deleted and
+	// its EndpointSlices are being rapidly updated.
+	// For more context: https://github.com/kubernetes/kubernetes/issues/91287
+	ginkgo.It("should not get in the way of service garbage collection", func() {
+		sharedLabels := map[string]string{"test-garbage-collection": "for-service"}
+
+		svc := createServiceReportErr(cs, f.Namespace.Name, &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "example-int-port",
+			},
+			Spec: v1.ServiceSpec{
+				// No need to wait for pods to be ready for this test
+				PublishNotReadyAddresses: true,
+				Selector:                 sharedLabels,
+				Ports: []v1.ServicePort{{
+					Name:       "example",
+					Port:       80,
+					TargetPort: intstr.FromInt(3000),
+					Protocol:   v1.ProtocolTCP,
+				}},
+			},
+		})
+
+		// Create an initial set of 100 test pods for initial EndpointSlice.
+		createTestPods(podClient, 0, 100, sharedLabels)
+
+		// Ensure EndpointSlice has been created for Service and has at least 50
+		// endpoints. Not waiting for all Pods to be part of EndpointSlice to
+		// ensure EndpointSlice controller is even busier in following step.
+		lSelector := discoveryv1beta1.LabelServiceName + "=" + svc.Name
+		err := wait.Poll(5*time.Second, 3*time.Minute, func() (bool, error) {
+			esList, err := cs.DiscoveryV1beta1().EndpointSlices(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
+			framework.ExpectNoError(err, "Error listing EndpointSlices for %s/%s Service", f.Namespace.Name, svc.Name)
+			if len(esList.Items) != 1 {
+				framework.Logf("Expected 1 EndpointSlice for %s/%s Service, got %d", f.Namespace.Name, svc.Name, len(esList.Items))
+				return false, nil
+			}
+
+			if len(esList.Items[0].Endpoints) < 50 {
+				framework.Logf("Expected at least 50 EndpointSlice endpoints for %s/%s Service, got %d", f.Namespace.Name, svc.Name, len(esList.Items[0].Endpoints))
+				return false, nil
+			}
+
+			return true, nil
+		})
+		framework.ExpectNoError(err, "Timed out waiting for EndpointSlice to have at least 50 endpoints for %s/%s Service", f.Namespace.Name, svc.Name)
+
+		// In this step we run 4 goroutines simultaneously. The first 3 are all
+		// trying to keep the EndpointSlice controller as busy as possible.
+		// There are 2 goroutines creating 51 new Pods each at the same time
+		// another goroutine is deleting the 100 Pods created initially. The
+		// final goroutine deletes the Service in the midst of the chaos to
+		// ensure that all of this action in the EndpointSlice controller will
+		// not interfere with garbage collection.
+		var wg sync.WaitGroup
+		wg.Add(4)
+
+		go func() {
+			defer wg.Done()
+			createTestPods(podClient, 100, 151, sharedLabels)
+		}()
+
+		go func() {
+			defer wg.Done()
+			createTestPods(podClient, 200, 251, sharedLabels)
+		}()
+
+		go func() {
+			defer wg.Done()
+			deleteTestPods(podClient, 0, 100, sharedLabels)
+		}()
```

Comment:
@liggitt Would the scale of this test be considered too disruptive? It's currently creating a total of 202 pods though less than 200 should exist at any one time. I was trying to find a way to simulate the EndpointSlice controller being busy without actually overloading the cluster with resources. This also has the benefit of exercising multiple EndpointSlices, something we'd need >100 pods for.

Reply:
hmm… our e2e struggles enough that I'd hesitate to intentionally add scaleish tests here. @wojtek-t , any thoughts?

Reply:
Yeah... Note that we're runnin 30 tests in parallel, and clusters we test it on have 5? 6? nodes. So it's pretty significant for them.

I'm not particularly afraid of running this test in large clusters, but for presubmit it may cause problems for random tests, which would make debugging it really hard.

Reply:
let's drop the load test from this and lean on the unit test.

Reply:
I talked with @wojtek-t about this and it sounds like it could be a good use of the `[Serial]` tag. I've added that for this test. I think it still provides significant value as the only e2e test to exercise multiple Endpoint Slices.